lib/actions/device_ts.ts

```diff
@@ -55,62 +56,67 @@ export const osUpdate: CommandDefinition<OsUpdate.Args, OsUpdate.Options> = {
 		const patterns = await import('../utils/patterns');
 		const form = await import('resin-cli-form');
 
-		return sdk.models.device
-			.get(params.uuid, {
-				$select: ['uuid', 'device_type', 'os_version', 'os_variant'],
-			})
-			.then(async ({ uuid, device_type, os_version, os_variant }) => {
-				const currentOsVersion = sdk.models.device.getOsVersion({
-					os_version,
-					os_variant,
-				} as Device);
-				if (!currentOsVersion) {
-					patterns.exitWithExpectedError(
-						'The current os version of the device is not available',
-					);
-					// Just to make TS happy
-					return;
-				}
-
-				return sdk.models.os
-					.getSupportedOsUpdateVersions(device_type, currentOsVersion)
-					.then(hupVersionInfo => {
-						if (hupVersionInfo.versions.length === 0) {
-							patterns.exitWithExpectedError(
-								'There are no available Host OS update targets for this device',
-							);
-						}
-
-						if (options.version != null) {
-							if (!_.includes(hupVersionInfo.versions, options.version)) {
-								patterns.exitWithExpectedError(
-									'The provided version is not in the Host OS update targets for this device',
-								);
-							}
-							return options.version;
-						}
-
-						return form.ask({
-							message: 'Target OS version',
-							type: 'list',
-							choices: hupVersionInfo.versions.map(version => ({
-								name:
-									hupVersionInfo.recommended === version
-										? `${version} (recommended)`
-										: version,
-								value: version,
-							})),
-						});
-					})
-					.then(version =>
-						patterns
-							.confirm(
-								options.yes || false,
-								'Host OS updates require a device restart when they complete. Are you sure you want to proceed?',
-							)
-							.then(() => sdk.models.device.startOsUpdate(uuid, version))
-							.then(() => patterns.awaitDeviceOsUpdate(uuid, version)),
-					);
+		// Get device info
+		const {
+			uuid,
+			device_type,
+			os_version,
+			os_variant,
+		} = await sdk.models.device.get(params.uuid, {
+			$select: ['uuid', 'device_type', 'os_version', 'os_variant'],
+		});
+
+		// Get current device OS version
+		const currentOsVersion = sdk.models.device.getOsVersion({
+			os_version,
+			os_variant,
+		} as Device);
+		if (!currentOsVersion) {
+			throw new ExpectedError(
+				'The current os version of the device is not available',
+			);
+		}
+
+		// Get supported OS update versions
+		const hupVersionInfo = await sdk.models.os.getSupportedOsUpdateVersions(
+			device_type,
+			currentOsVersion,
+		);
+		if (hupVersionInfo.versions.length === 0) {
+			throw new ExpectedError(
+				'There are no available Host OS update targets for this device',
+			);
+		}
+
+		// Get target OS version
+		let targetOsVersion = options.version;
+		if (targetOsVersion != null) {
```

Comment:
Hmm I not sure I'd agree that that would help with readability, I guess this is quite subjective.  If it were needed in more than one place then I'd agree that moving to a function is good from a modularity point of view, but as far as I know it's not.  Personally I think splitting simple code into functions can sometimes have the reverse effect.

Reply:
Subjective?! :-) Functions are great: they break large problems in smaller ones, define input/output contracts that help reason for correctness (when low coupling is possible), and facilitate unit testing (which in some companies is like religion). I am approving the PR anyway, but I see it a basic [software design](https://en.wikipedia.org/wiki/Software_design) principle/technique rather than my personal opinion.

Reply:
How about reformatting this to reduce nesting:
`if (targetOsVersion == null) {
  targetOsVersion = await form.ask({...});
}
// possibly make this an `else if` you prefer, but I do like the extra check anyway :)
if (!_.includes(hupVersionInfo.versions, targetOsVersion)) {
  throw...;
}